### DISCUSSION TITLE
What's the diffence between the network topology and the wastewater network elements connections

### DISCUSSION THREAD

**Discussion by rbovard** [QGEP/QGEP#758] [Q&A] [answered]
What's the diffence between the **network topology** and the **wastewater network elements connections**?

Because I can generate the network topology with the `Refresh network topology` tool but I need to manually connect reaches and wastewate structure manually with the `Connect wastewater network elements`.

I think I'm missing something obvious here...

**Comment by urskaufmann** [comment, marked as answer]
Refresh network topology creates/refreshes the two layers, that are necessary for network-following and length profil. It uses only the connections in the foreignkey-fields. The position of the lines and points has no effect. It does not change any foreignkey value.

With the tool "Connect wastewater network elements" you define the values of these foreignkey-fields.

**Comment by rbovard** [comment]
Thanks @urskaufmann for this answer.

> It uses only the connections in the foreignkey-fields.

I do not agree, because when running the refresh network tool without any foreign keys, it generates me the graph based on the position of reaches and structures.

This is why I'm confused about that...

**Reply by sjib** [reply]
@rbovard Can you share a screenshot?

**Reply by rbovard** [reply]
The reach without any start/end nodes:
![image](https://user-images.githubusercontent.com/4549666/194534811-77953f01-1deb-4e3f-bc6c-5e17c0c90501.png)

The generated network segment:
![image](https://user-images.githubusercontent.com/4549666/194535513-118cb008-54d3-4c99-aa96-53ecf65a503d.png)

The reach when I use the `Connect wastewater network elements` tool:
![image](https://user-images.githubusercontent.com/4549666/194535890-eb86fdc5-f7ed-43d6-b527-1ada822d3838.png)

And the result when I use the `Refresh network topology` tool:
![image](https://user-images.githubusercontent.com/4549666/194536315-73db4b8a-9d6b-401f-8a62-e63f30f056c6.png)

![image](https://user-images.githubusercontent.com/4549666/194536759-b3cd6e13-bbfd-40c7-a5e8-954a43e59faf.png)

**Reply by sjib** [reply]
@rbovard The last image is strange: Why do you have two arrows in the middle? It looks like the water would flow in that pipe from the left AND the right side. The second image looked much more as I would expect it to look like.

Did you only make the network connections from WS <- WN <- RP from <- RE -> RP to -> WN -> WS in one channel / reach? 
The one you display? Or is this a channel with two reaches?

**Reply by rbovard** [reply]
Yes it's very strange, it's the result I had after manually connect the reach to the two structures and regenerating the network. Maybe something will ring a bell on your side when looking at the `(from|to)_obj_id` attributes in my identify result?

Yes, I only have one reach, which generated the 3 network segments...

**Reply by rbovard** [reply]
Here is my workflow:
![topology](https://user-images.githubusercontent.com/4549666/194565579-d9a22eb9-37a2-4fe7-8ee3-80d007b56495.gif)

Am I doing something wrong?

**Reply by sjib** [reply]
I would rather choose the other way round: Left side node in the picture to node, right side from node (would for me be more logical with the rest of the network).

Can you try to start from the beginning of  branch with 2, 3 pipes down (and not inmidst the whole network)?

**Comment by urskaufmann** [comment]
for me its quite clear: when defining the connection, you have on both side changed the checkboxes from "to" to "from" and from "from" to "to". You have not connected the node next to reachpoint, but the node on the other side of the reach. Therefor you have then 3 elements in network-segement. 1 reach and 2 reachpoint-to-node connections....

**Comment by urskaufmann** [comment]
As I did write above: you will never have to change this checkboxes, if you start with the connection on the reach next to the end of the reach, where you wnat to connect.
